_3llo: init at 0.3.0

Ma27 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Simple CLI client for `trello.com`. It can be used like this:

```
$ export TRELLO_USER=your_username
$ export TRELLO_KEY=your_key
$ export TRELLO_TOKEN=your_token
$ ./result/bin/3llo
```

I didn't create a module for this as I don't think that those secrets
should live in the Nix store. Ideally `3llo` can be used from a script
which retrieves secrets from some kind of password store like this:

```
export TRELLO_KEY=$(pass show trello/key)
export TRELLO_TOKEN=$(pass show trello/token)
3llo $@
```



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ma27 committed:
_3llo: init at 0.3.0

Simple CLI client for `trello.com`. It can be used like this:

```
$ export TRELLO_USER=your_username
$ export TRELLO_KEY=your_key
$ export TRELLO_TOKEN=your_token
$ ./result/bin/3llo
```

I didn't create a module for this as I don't think that those secrets
should live in the Nix store. Ideally `3llo` can be used from a script
which retrieves secrets from some kind of password store like this:

```
export TRELLO_KEY=$(pass show trello/key)
export TRELLO_TOKEN=$(pass show trello/token)
3llo $@
```

worldofpeace commented:
Why is there an underscore in the attribute name when the repo doesn't name the tool that way?
* https://github.com/qcam/3llo

worldofpeace commented on pkgs/top-level/all-packages.nix:
```diff
@@ -435,6 +435,8 @@ in
 
   _0x0 = callPackage ../tools/misc/0x0 { };
 
+  _3llo = callPackage ../applications/misc/3llo { };
```
Would this be a better fit for `tools` since it's such a small program?

##### also note this is already in the `### TOOLS` section and not `### APPLICATIONS`

worldofpeace reviewed (changes requested):
Expression looks fine to me.

Not sure about its place in nixpkgs though.

worldofpeace commented:
It would be nice if this program self documented itself instead of erroring ungracefully when you don't set the aforementioned environment variables

```
❯ ./result/bin/3llo --help
Traceback (most recent call last):
	4: from ./result/bin/3llo:18:in `<main>'
	3: from ./result/bin/3llo:18:in `load'
	2: from /nix/store/dsqvych6dja9rvb2ar4fj914zhlwykl4-3llo-0.3.0/lib/ruby/gems/2.5.0/gems/3llo-0.3.0/bin/3llo:14:in `<top (required)>'
	1: from /nix/store/dsqvych6dja9rvb2ar4fj914zhlwykl4-3llo-0.3.0/lib/ruby/gems/2.5.0/gems/3llo-0.3.0/bin/3llo:14:in `fetch'
/nix/store/dsqvych6dja9rvb2ar4fj914zhlwykl4-3llo-0.3.0/lib/ruby/gems/2.5.0/gems/3llo-0.3.0/bin/3llo:14:in `block in <top (required)>': Have you set TRELLO_USER? (RuntimeError)
```

Ma27 commented:
> Why is there an underscore in the attribute name when the repo doesn't name the tool that way?

Nix doesn't like attribute names with a number at first. I was unsure how to deal with this and I just used this pattern from other tools like `_1password`.

> It would be nice if this program self documented itself instead of erroring ungracefully when you don't set the aforementioned environment variables

In fact you're right. Not sure if I can make it today, but I'll prepare an upstream patch for this (unless there's something that just didn't land in the release :) )

> Would this be a better fit for tools since it's such a small program?

Hmm, probably yes.... I'd like to note though that I actually dislike this section-based approach as it makes it IMHO way harder to find an appropriate place for a new package and I honestly doubt that many people benefit from this... at least whenever I look for something I either grep over `nixpkgs` or `all-packages.nix`.

worldofpeace commented:
> > Why is there an underscore in the attribute name when the repo doesn't name the tool that way?
>
> Nix doesn't like attribute names with a number at first. I was unsure how to deal with this and I just used this pattern from other tools like _1password.

I think we can just make it `"3llo"` in `all-packages.nix`.

> > It would be nice if this program self documented itself instead of erroring ungracefully when you don't set the aforementioned environment variables
>
> In fact you're right. Not sure if I can make it today, but I'll prepare an upstream patch for this (unless there's something that just didn't land in the release :) )

That would be really excellent if you did :sparkles: 

> > Would this be a better fit for tools since it's such a small program?
> Hmm, probably yes.... I'd like to note though that I actually dislike this section-based approach as it makes it IMHO way harder to find an appropriate place for a new package and I honestly doubt that many people benefit from this... at least whenever I look for something I either grep over nixpkgs or all-packages.nix.

I think it helps with breaking up the psuedo alphabetization. Then at least you can see it restarting
at least in definite sections.

Ma27 commented:
> I think we can just make it "3llo" in all-packages.nix.

But isn't that counter-intuitive when you want to add a bunch of packages like `with pkgs; [ foo bar ]` where you'd have to reference `3llo` with `pkgs."3llo"`? (I'm surprised that `nix-build -A 3llo` would work btw).

worldofpeace commented:
> > I think we can just make it "3llo" in all-packages.nix.
> 
> But isn't that counter-intuitive when you want to add a bunch of packages like `with pkgs; [ foo bar ]` where you'd have to reference `3llo` with `pkgs."3llo"`? (I'm surprised that `nix-build -A 3llo` would work btw).

Huh, we would then get a syntax error when referencing it in the function arguments of a package
```
syntax error, unexpected '"', expecting '}'
```

I don't think there's a guideline in the Nixpkgs manual about this, rather unfortunate that it would have to be `_3llo`.

Ma27 committed:
ruby/bundlerApp: ignore `gemset` in env attribute set

The `gemset` field can be an attribute set as well in `buildRubyGem`,
however attribute sets can't be coerced to strings for the environment
set.

The impact should be relatively small as the environment variables are
only used by the `runCommand` script for `bunlderApp` which doesn't
refer to `gemset` at all.

Ma27 commented:
@worldofpeace updated. Would be awesome if you could review the upstream patch as well. Please keep in mind that I had to change a minor detail in `bundlerApp`. I'm currently running `nix-review`, but I definetely need more reviewers then :)

worldofpeace commented:
Gave it a review @Ma27

Ma27 commented:
@worldofpeace updated the patch :)

worldofpeace reviewed (approved):
LGTM, thanks @Ma27.
